presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java

```diff
@@ -1391,14 +1372,14 @@ private void printPlanNodesStatsAndCost(int indent, PlanNode... nodes)
 
         private boolean isKnownPlanNodeStatsOrCost(PlanNode node)
         {
-            return !UNKNOWN_STATS.equals(statsProvider.getStats(node))
-                    || !UNKNOWN_COST.equals(costProvider.getCumulativeCost(node));
+            return !UNKNOWN_STATS.equals(estimatedStatsAndCosts.getStats(node))
```

Comment:
I would refactor this part to be like this

```
private void printPlanNodesStatsAndCost(int indent, PlanNode... nodes)
        {
            List<String> statsAndCosts = Arrays.stream(nodes)
                    .map(this::formatPlanNodeStatsAndCost)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .collect(toImmutableList());

            if (!statsAndCosts.isEmpty()) {
                print(indent, "Cost: %s", Joiner.on("/").join(statsAndCosts));
            }
        }

        private Optional<String> formatPlanNodeStatsAndCost(PlanNode node)
        {
            PlanNodeStatsEstimate stats = estimatedStatsAndCosts.getStats().get(node.getId());
            PlanNodeCostEstimate cost = estimatedStatsAndCosts.getCosts().get(node.getId());
            if (stats == null || cost == null) {
                return Optional.empty();
            }
            return Optional.of(String.format("{rows: %s (%s), cpu: %s, memory: %s, network: %s}",
                    formatAsLong(stats.getOutputRowCount()),
                    formatEstimateAsDataSize(stats.getOutputSizeInBytes(node.getOutputSymbols(), types)),
                    formatDouble(cost.getCpuCost()),
                    formatDouble(cost.getMemoryCost()),
                    formatDouble(cost.getNetworkCost())));
        }
```

Also consider not printing stats if stats for some node is missing. If, for example, for `ScanFilterAndProject`, the stats are missing for `Filter` - the output may be confusing.

Reply:
I like that.  I'm doing that with a slight modification that I'll still check for unknown because it's still possible for the Stats/Cost maps to contain UNKNOWN_STATS and I don't think we gain anything from enforcing that in StatsAndCosts itself.